database update improvements

Pseudodude opened the pull request:
- Avoid commits during update to improve performance
- Update from AllCards.json.zip instead of plaintext AllCards.json

Pseudodude committed:
database update improvements

- Avoid commits during update to improve performance
- Update from AllCards.json.zip instead of plaintext AllCards.json

bakert commented on database.py:
```diff
@@ -29,11 +29,13 @@ def execute(self, sql, args = []):
-    try:
-      return rs[0][0]
-    except IndexError:
+    rs = self.database.execute(sql, args).fetchone()
```
rs is short for "resultset" or "recordset". If we're only fetching a single row then 'row' would be a more appropriate name.

bakert commented on database.py:
```diff
@@ -29,11 +29,13 @@ def execute(self, sql, args = []):
+    if rs is None:
+      return default
+    elif len(rs) <= 0:
       return default
```
if rs is None or len(rs) == 0:
  return default

?

bakert commented on fetcher.py:
```diff
@@ -10,7 +10,17 @@ def version(self):
 
   def all_cards(self):
-    return json.loads(self.open('https://mtgjson.com/json/AllCards.json'))
+    if os.path.isdir('./ziptemp'):
```
We let the user choose locations for db and images. I wonder if we should just have single entry in config.json for "directory to store stuff in" and put this in it too.

Then we can not delete it afterwards and check its timestamp before bothering to re-download it (if it's less than hour old, say).

bakert commented on fetcher.py:
```diff
@@ -10,7 +10,17 @@ def version(self):
+      shutil.rmtree('./ziptemp')
+      
+    os.mkdir('./ziptemp')
+    urllib.request.urlretrieve('https://mtgjson.com/json/AllCards.json.zip', './ziptemp/AllCards.json.zip')
```
We should really make "download {url} to {location}" a method of fetcher and call it when fetching images from bluebones or gatherer, too.

bakert commented on oracle.py:
```diff
@@ -73,20 +75,22 @@ def insert_card(self, name, card):
     id = self.database.value('SELECT last_insert_rowid()')
     for name in card.get('names', []):
-      self.database.execute('INSERT INTO card_name (card_id, name) VALUES (?, ?)', [id, name])
+      self.database.database.execute('INSERT INTO card_name (card_id, name) VALUES (?, ?)', [id, name])
```
Good God do I hate self.database.database :)

I feel like we're reaching into the internals of Database here. It might be nicer to build up an array of arrays of args and then process them all at the end? sqlite3 exposes executeMany which may or may not be useful.

If that's too much surgery then perhaps we can just expose a version of execute that doesn't commit or have a boolean param on execute for commit-or-not?

Pseudodude replied:
self.database.database could become self.database.connection, but I take your point about calling the function on Database's instance variable. I didn't want to change or rename the current execute+commit of my own accord, but adding a boolean param sounds like the best solution.

bakert commented on database.py:
```diff
@@ -99,5 +101,5 @@ def setup(self):
     """)
 
-# Import last to work around circular dependency — http://effbot.org/zone/import-confusion.htm
+# Import last to work around circular dependency � http://effbot.org/zone/import-confusion.htm
```
What changed here? Trailing whitespace? In a good way or a bad way?

bakert reviewed (approved):
Some things to think about but nothing blocking.

Pseudodude committed:
Bug fix: AllCards.json should be loaded with utf-8 encoding after unzipping

Was being loaded with default encoding, so cards with accents in their
name were being given incorrect names in the db.
